optimade/validator/utils.py

```diff
@@ -142,23 +138,119 @@ def add_failure(
         elif failure_type == "optional":
             self.optional_failure_count += 1
             self.optional_failure_messages.append((summary, message))
+        elif failure_type == "warning":
+            self.warning_count += 1
+            self.warning_messages.append((summary, message))
+
+        self._print_test_failure(summary, message, failure_type=failure_type)
+
+    def _print_test_success(
+        self, summary: str, success_type: str = None, verbosity: Optional[int] = None
+    ):
+        """Verbosity-dependent printing of a single test case success message.
+
+        Parameters:
+            summary: The message to display.
+            success_type: Either `None` or `"optional"`.
+            verbosity: If 0, do not print the summary, defaults to stored value.
+
+        """
+        if verbosity is None:
+            verbosity = self.verbosity
+
+        message = f"✔: {summary}"
+        pretty_print = print if success_type == "optional" else print_success
+
+        if self.verbosity > 0:
+            pretty_print(message)
+        elif self.verbosity == 0:
+            pretty_print(".", end="", flush=True)
+
+    def _print_test_failure(self, summary, message, failure_type=None, verbosity=None):
+        """Verbosity-dependent printing of a single test case failure message.
+
+        Parameters:
+            summary: The summary message to display.
+            message: The full error description/traceback.
+            failure_type: Either `None` or `"optional"`, `"internal"`
+                or `"warning"`.
+            verbosity: If 0, do not print the summary, defaults to
+                stored value.
 
+        """
+        if verbosity is None:
+            verbosity = self.verbosity
         pprint_types = {
             "internal": (print_notify, print_warning),
             "optional": (print, print),
+            "warning": (print_notify, print),
         }
         pprint, warning_pprint = pprint_types.get(
             failure_type, (print_failure, print_warning)
         )
 
-        symbol = "!" if failure_type == "internal" else "✖"
+        symbols = {"internal": "!", "warning": "~", "optional": "✖"}
```

Comment:
Should this be a general attribute somewhere instead?

Reply:
Potentially, but its only used here

Reply:
Fair enough. I would probably still expect this to be some general resource, but if it's only used here. No worries.